cordapp-upgrades/workflows/v1-initial/src/main/kotlin/net/corda/examples/obligation/workflow/flows/SettleObligation.kt

```diff
@@ -0,0 +1,125 @@
+package net.corda.examples.obligation.workflow.flows
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.confidential.IdentitySyncFlow
+import net.corda.core.contracts.Amount
+import net.corda.core.contracts.Command
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.flows.*
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.utilities.ProgressTracker
+import net.corda.examples.obligation.contract.ObligationContract
+import net.corda.examples.obligation.contract.ObligationContract.Companion.OBLIGATION_CONTRACT_ID
+import net.corda.finance.contracts.asset.PartyAndAmount
+import net.corda.finance.workflows.asset.CashUtils
+import net.corda.finance.workflows.getCashBalance
+import java.util.*
+
+object SettleObligation {
+    @InitiatingFlow
+    @StartableByRPC
+    class Initiator(private val linearId: UniqueIdentifier,
+                    private val amount: Amount<Currency>,
+                    private val anonymous: Boolean = true) : ObligationBaseFlow() {
+
+        override val progressTracker: ProgressTracker = tracker()
+
+        companion object {
+            object PREPARATION : ProgressTracker.Step("Obtaining Obligation from vault.")
+            object BUILDING : ProgressTracker.Step("Building and verifying transaction.")
+            object SIGNING : ProgressTracker.Step("signing transaction.")
+            object COLLECTING : ProgressTracker.Step("Collecting counterparty signature.") {
+                override fun childProgressTracker() = CollectSignaturesFlow.tracker()
+            }
+            object FINALISING : ProgressTracker.Step("Finalising transaction.") {
+                override fun childProgressTracker() = FinalityFlow.tracker()
+            }
+
+            fun tracker() = ProgressTracker(PREPARATION, BUILDING, SIGNING, COLLECTING, FINALISING)
+        }
+
+        @Suspendable
+        override fun call(): SignedTransaction {
+            // Stage 1. Retrieve obligation specified by linearId from the vault.
+            progressTracker.currentStep = PREPARATION
+            val obligationToSettle = getObligationByLinearId(linearId)
+            val inputObligation = obligationToSettle.state.data
+
+            // Stage 2. Resolve the lender and borrower identity if the obligation is anonymous.
+            val borrowerIdentity = resolveIdentity(inputObligation.borrower)
+            val lenderIdentity = resolveIdentity(inputObligation.lender)
+
+            // Stage 3. This flow can only be initiated by the current recipient.
+            check(borrowerIdentity == ourIdentity) {
+                throw FlowException("Settle Obligation flow must be initiated by the borrower.")
```

Comment:
This is completely wrong. `check` throws an IllegalStateException if the boolean expression false. The lambda is here to provide a message string for that exception, not to throw another exception! I presume this is how it is in the original CorDapp. If so can you go back and apply the same changes in a separate PR. We probably need our own version of `check` that throws FlowException.

Reply:
It feels to me like a version of `check` that throws `FlowException` would be generally useful - perhaps something to add to the core of the platform?